### PR TITLE
feat: Adding "module" field to logs with module of actual logger

### PR DESCRIPTION
**Pull request by deregtd**
Walking stack with caching stack frame lookups to find out who is calling the logger and adding it to all formatted logs.  Intent is to be able to filter out what application is the actual source of logs for filtering/cost attribution.

### CODE REVIEW

File: pkg/callerinfo/callerinfo.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		// Found it in the cache
+		return mod, nil
+	}
+
+	// Not cached -- have to do the slow lookup
+	frames := runtime.CallersFrames(pc)
+	frame, _ := frames.Next()
+
+	// Cache for later
+	moduleLookupByPC[pc[0]] = frame.Function
```
Comment (jkinkead) [inline pkg/callerinfo/callerinfo.go:40]: You need a lock around the write here.

File: pkg/log/log.go [hunk 2]
```diff
@@ -150,6 +151,29 @@ func format(msg, level string, ts time.Time, appInfo Marshaler, mm Many) string
...
+	caller, err := callerinfo.GetCallerFunction(2)
+	switch {
+	case err != nil:
+		entry["source"] = "error"
+	case strings.HasPrefix(caller, "github.com/"):
+		// Example response: github.com/getoutreach/gobox/pkg/callerinfo.Test_Callers
+		splits := strings.Split(caller, "/")
+		// We're going for extracting "getoutreach/gobox" in the above example
+		if len(splits) >= 3 {
+			entry["source"] = splits[1] + "/" + splits[2]
```
Comment (jkinkead) [inline pkg/log/log.go:165]: I would not strip off `github.com`, as it's part of the module name.

File: pkg/log/log.go [hunk 3]
```diff
@@ -150,6 +151,29 @@ func format(msg, level string, ts time.Time, appInfo Marshaler, mm Many) string
...
+		} else {
+			entry["source"] = splits[1]
+		}
+	case strings.HasPrefix(caller, "main"):
+		// Main entry point for the app shows up as "main.[funcname]", so just use the app name since we unfortunately
+		// don't have a full package name
+		entry["source"] = app.Info().Name
```
Comment (jkinkead) [inline pkg/log/log.go:172]: Prefix with `app:` to help disambiguate?

Comment (deregtd) [reply pkg/log/log.go:172]: i like that, will do

File: pkg/log/log.go (hunk 2 above)
Comment (jaredallard) [reply pkg/log/log.go:165]: I really wish we had a way to get the canonical module name from https://pkg.go.dev/runtime/debug#ReadBuildInfo...

File: pkg/log/log.go (hunk 3 above)
Comment (jaredallard) [reply pkg/log/log.go:172]: Or should `source` be the module name as it appears to Go? That goes back to my earlier comment about https://pkg.go.dev/runtime/debug#ReadBuildInfo.

File: pkg/callerinfo/callerinfo.go (hunk 1 above)
Comment (jaredallard) [reply pkg/callerinfo/callerinfo.go:40]: Should we consider having this be an LRU cache?

File: pkg/log/log.go (hunk 2 above)
Comment (jaredallard) [reply pkg/log/log.go:165]: Noted this in a DM as well, but we could use the aforementioned build information to retrieve the version of the library in use and store it in a separate field like `source_version`.

File: pkg/callerinfo/callerinfo.go (hunk 1 above)
Comment (deregtd) [reply pkg/callerinfo/callerinfo.go:40]: @jkinkead oh goddamnit, I thought it was both read and write concurrency safe, but it's only read.  Sigh.  Thanks.

@jaredallard It's capped to a count of all of the locations in the app making a log send, so I didn't feel like it made sense to add all the overhead of an LRU cache here.  Even if you have an outstandingly loud app with thousands of log locations, it's still going to be, worst case, like 100-200k of memory.

Comment (jaredallard) [reply pkg/callerinfo/callerinfo.go:40]: Makes sense, just wanted to double-check that. I figured I was trying to overoptimize here :D

File: pkg/log/log.go [hunk 4]
```diff
@@ -150,6 +151,29 @@ func format(msg, level string, ts time.Time, appInfo Marshaler, mm Many) string
 		generateFatalFields(entry)
 	}
 
+	// Attempt to map the caller of the log function into the "source" field for identifying if a service or a module
```
Comment (jaredallard) [inline pkg/log/log.go:154]: nit: I'd suggest pulling this out into its own function so we reduce the bloat of this function. That being said, I don't really care about this logger at all and want to kill it with fire as soon as possible so I am happy to ignore this

Comment (deregtd) [reply pkg/log/log.go:154]: ya totes fair

File: pkg/log/log.go (hunk 2 above)
Comment (deregtd) [reply pkg/log/log.go]: done -- reworked the whole thing

**Review by jkinkead — APPROVED** [review]
LGTM although we might be building in some custom labeling we'll need to retrofit into another solution like logrus.

File: pkg/callerinfo/callerinfo.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+package callerinfo
+
+import (
+	"fmt"
+	"runtime"
+	"runtime/debug"
+	"strings"
+	"sync"
+)
+
```
Comment (jaredallard) [inline pkg/callerinfo/callerinfo.go:16]: exported, added comment :D

Comment (deregtd) [reply pkg/callerinfo/callerinfo.go:16]: oh, rip good clal

File: pkg/callerinfo/callerinfo.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	frame, _ := frames.Next()
+
+	ci := CallerInfo{
+		Function: frame.Function,
+		File:     frame.File,
+		LineNum:  uint(frame.Line),
+	}
+
+	// Attempt to back-calc module name by searching through deps
+	if buildInfo != nil {
```
Comment (jaredallard) [inline pkg/callerinfo/callerinfo.go:67]: def move this into another function, having it be nested makes it harder to read here to begin w/ but also it fits into it's own function p easily

File: pkg/callerinfo/callerinfo.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+		for _, mod := range buildInfo.Deps {
+			if strings.HasPrefix(ci.Function, mod.Path) {
+				ci.Module = mod.Path
+				ci.ModuleVersion = mod.Version
+				break
+			}
+		}
+
+		// If we can't find it in the dep list, it must be from the main app
+		if ci.Module == "" {
```
Comment (jaredallard) [inline pkg/callerinfo/callerinfo.go:77]: This seems like an unsafe assumption, maybe we could at least see if `Main.Path` is in the string?

Comment (deregtd) [reply pkg/callerinfo/callerinfo.go:77]: It's been true in all my circumstances.  the function shows up as, for example, "main.main", so there's no path prefix :(

Comment (jaredallard) [reply pkg/callerinfo/callerinfo.go:77]: Lame, ok. I guess we gotta :(

Can fix it later p easily I guess

**Review by jaredallard — APPROVED** [review]

File: pkg/callerinfo/callerinfo.go (hunk 7 above)
Comment (deregtd) [reply pkg/callerinfo/callerinfo.go:77]: yeah, we'll see how it looks when we go wider.  this is all heuristicy :/

File: pkg/callerinfo/callerinfo.go (hunk 6 above)
Comment (deregtd) [reply pkg/callerinfo/callerinfo.go:75]: okie done

**Pull request merged**